server/blob/blob-objectstorage/src/main/java/org/apache/james/blob/objectstorage/aws/AwsS3ObjectStorage.java

```diff
@@ -19,27 +19,130 @@
+    private static final int MAX_UPLOAD_THREADS = 5;
+    private static final int MAX_ERROR_RETRY = 5;
+    private static final int FIRST_TRY = 0;
+    private static final int MAX_RETRY_ON_EXCEPTION = 3;
+    private static Size MULTIPART_UPLOAD_THRESHOLD;
+
+    static {
+        try {
+            MULTIPART_UPLOAD_THRESHOLD = Size.parse("5M");
+        } catch (Exception e) {
+            throw new RuntimeException(e);
+        }
+    }
 
     public static ObjectStorageBlobsDAOBuilder.RequireContainerName daoBuilder(AwsS3AuthConfiguration configuration) {
         return ObjectStorageBlobsDAOBuilder.forBlobStore(new BlobStoreBuilder(configuration));
     }
 
+    public static Optional<PutBlobFunction> putBlob(ContainerName containerName, AwsS3AuthConfiguration configuration) {
+        return Optional.of((blob) -> {
+            File file = null;
+            try {
+                file = File.createTempFile(UUID.randomUUID().toString(), ".tmp");
+                FileUtils.copyToFile(blob.getPayload().openStream(), file);
+                putWithRetry(containerName, configuration, blob, file, FIRST_TRY);
+            } catch (IOException e) {
+                throw new RuntimeException(e);
+            } finally {
+                if (file != null) {
+                    FileUtils.deleteQuietly(file);
+                }
+            }
+        });
+    }
+
+    private static void putWithRetry(ContainerName containerName, AwsS3AuthConfiguration configuration, Blob blob, File file, int tried) {
+        try {
+            put(containerName, configuration, blob, file);
+        } catch (RuntimeException e) {
+            if (tried < MAX_RETRY_ON_EXCEPTION) {
+                putWithRetry(containerName, configuration, blob, file, tried + 1);
+            } else {
+                throw e;
+            }
+        }
+    }
+
+    private static void put(ContainerName containerName, AwsS3AuthConfiguration configuration, Blob blob, File file) {
+        try {
+            PutObjectRequest request = new PutObjectRequest(containerName.value(),
+                blob.getMetadata().getName(),
+                file);
+
+            getTransferManager(configuration)
+                .upload(request)
+                .waitForUploadResult();
+        } catch (AmazonClientException | InterruptedException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    private static TransferManager getTransferManager(AwsS3AuthConfiguration configuration) {
+        ClientConfiguration clientConfiguration = getClientConfiguration();
+        AmazonS3 amazonS3 = getS3Client(configuration, clientConfiguration);
+
+        return TransferManagerBuilder
+            .standard()
+            .withS3Client(amazonS3)
+            .withMultipartUploadThreshold(MULTIPART_UPLOAD_THRESHOLD.getValue())
+            .withExecutorFactory(() -> Executors.newFixedThreadPool(MAX_UPLOAD_THREADS))
```

Comment:
I think I already asked for a named threadpool and this threadpool should be created only once.

Reply:
Oh, and the threadpool need to be shutdown at some point

Reply:
I removed this 'fix' on purpose as the s3 client kill the executor after an upload is complete.
i will have a second look at it, as there is an option to avoid this behavior.